modules/battery-udev.c

```diff
@@ -1776,6 +1794,131 @@ udevdevice_evaluate_battery_cb(gpointer key, gpointer value, gpointer aptr)
     udevdevice_evaluate_battery(self, mcebat);
 }
 
+/* ========================================================================= *
+ * UDEVEXTCON
+ * ========================================================================= */
+
+/** Subsystem name for extcon devices */
+static const char udevextcon_subsystem[] = "extcon";
+
+/** USB state cache for extcon devices */
+static GHashTable *udevextcon_usb_state_lut = NULL;
+
+/** Parse USB entry from extcon state information
+ *
+ * When non-null value is returned, caller must release it with g_free().
+ *
+ * @param state Full state data for extcon device
+ *
+ * @return value for USB entry, or NULL
+ */
+static gchar *
+udevextcon_parse_usb_state(const char *state)
+{
+    gchar *res = NULL;
+    gchar *tmp = NULL;
+
+    if( (tmp = g_strdup(state)) ) {
+        for( char *pos = tmp; *pos; ) {
+            char *val = mce_slice_token(pos, &pos, NULL);
+            char *key = mce_slice_token(val, &val, "=");
+            if( !strcmp(key, "USB") ) {
+                res = g_strdup(val);
+                break;
+            }
+        }
+        g_free(tmp);
+    }
+
+    return res;
+}
+
+/** Update cached USB state for an extcon device
+ *
+ * @param syspath  syspath for device to update
+ * @param state    state reported for the device
+ *
+ * @return true if cached value changed, false otherwise
+ */
+static bool
+udevextcon_update_state(const char *syspath, const char *state)
+{
+    bool         changed   = false;
+    gchar       *usb_state = NULL;
+    const gchar *old_state = NULL;
+
+    if( !syspath || !udevextcon_usb_state_lut )
+        goto EXIT;
+
+    usb_state = udevextcon_parse_usb_state(state);
+    old_state = g_hash_table_lookup(udevextcon_usb_state_lut, syspath);
+    if( g_strcmp0(old_state, usb_state) ) {
+        mce_log(LL_DEBUG, "%s.STATE / USB: %s -> %s",
+                basename(syspath), old_state ?: "null", usb_state ?: "null");
+        if( usb_state )
+            g_hash_table_replace(udevextcon_usb_state_lut, g_strdup(syspath),
+                                 usb_state), usb_state = NULL;
+        else
+            g_hash_table_remove(udevextcon_usb_state_lut, syspath);
+        changed = true;
+    }
+
+EXIT:
+    g_free(usb_state);
```

Comment:
Isn't this only needed before EXIT, since when goto EXIT is used usb_state is always NULL?

Reply:
The "pattern" is:  prioritize clarity of dynamic resource management over possible micro-optimizations -> if it is possible that a resource can get allocated, make it obvious that it will be released -> null init at entry, unskippable release before leaving.